### Вручтель Серафима. 495 группа. Наивный байесовский классификатор.

In [2]:
from sklearn import datasets

import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

Загрузим датасеты digits и breast_cancer из sklearn.datasets.

In [4]:
digits = datasets.load_digits()
breast_cancer = datasets.load_breast_cancer()


Выведем несколько строчек из обучающих выборок и посмотрим на признаки.

In [8]:
digits.keys()

dict_keys(['data', 'target', 'target_names', 'images', 'DESCR'])

In [13]:
digits.data[:2]

array([[  0.,   0.,   5.,  13.,   9.,   1.,   0.,   0.,   0.,   0.,  13.,
         15.,  10.,  15.,   5.,   0.,   0.,   3.,  15.,   2.,   0.,  11.,
          8.,   0.,   0.,   4.,  12.,   0.,   0.,   8.,   8.,   0.,   0.,
          5.,   8.,   0.,   0.,   9.,   8.,   0.,   0.,   4.,  11.,   0.,
          1.,  12.,   7.,   0.,   0.,   2.,  14.,   5.,  10.,  12.,   0.,
          0.,   0.,   0.,   6.,  13.,  10.,   0.,   0.,   0.],
       [  0.,   0.,   0.,  12.,  13.,   5.,   0.,   0.,   0.,   0.,   0.,
         11.,  16.,   9.,   0.,   0.,   0.,   0.,   3.,  15.,  16.,   6.,
          0.,   0.,   0.,   7.,  15.,  16.,  16.,   2.,   0.,   0.,   0.,
          0.,   1.,  16.,  16.,   3.,   0.,   0.,   0.,   0.,   1.,  16.,
         16.,   6.,   0.,   0.,   0.,   0.,   1.,  16.,  16.,   6.,   0.,
          0.,   0.,   0.,   0.,  11.,  16.,  10.,   0.,   0.]])

In [14]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [18]:
digits.target[:15]

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4])

In [24]:
print("keys:", breast_cancer.keys())
print("data[:2]:", breast_cancer.data[:2])
print("target_names:", breast_cancer.target_names)
print("target[:25]:", breast_cancer.target[:25])


keys: dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])
data[:2]: [[  1.79900000e+01   1.03800000e+01   1.22800000e+02   1.00100000e+03
    1.18400000e-01   2.77600000e-01   3.00100000e-01   1.47100000e-01
    2.41900000e-01   7.87100000e-02   1.09500000e+00   9.05300000e-01
    8.58900000e+00   1.53400000e+02   6.39900000e-03   4.90400000e-02
    5.37300000e-02   1.58700000e-02   3.00300000e-02   6.19300000e-03
    2.53800000e+01   1.73300000e+01   1.84600000e+02   2.01900000e+03
    1.62200000e-01   6.65600000e-01   7.11900000e-01   2.65400000e-01
    4.60100000e-01   1.18900000e-01]
 [  2.05700000e+01   1.77700000e+01   1.32900000e+02   1.32600000e+03
    8.47400000e-02   7.86400000e-02   8.69000000e-02   7.01700000e-02
    1.81200000e-01   5.66700000e-02   5.43500000e-01   7.33900000e-01
    3.39800000e+00   7.40800000e+01   5.22500000e-03   1.30800000e-02
    1.86000000e-02   1.34000000e-02   1.38900000e-02   3.53200000e-03
    2.49900000e+01   2.34100000e+01

In [25]:
# malignant означает "злокачественный"
# benign - "доброкачественный"


С помощью sklearn.model_selection.cross_val_score c настройками по умолчанию и вызова метода mean() у возвращаемого этой функцией numpy.ndarray, сравним качество работы наивных байесовских классификаторов на этих двух датасетах.

In [27]:
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB
from sklearn.model_selection import cross_val_score


In [31]:
X_d = digits.data
y_d = digits.target

X_bc = breast_cancer.data
y_bc = breast_cancer.target

# Создаём Gaussian Classifier
model_gaussian = GaussianNB()

# Создаём Bernoulli Classifier
model_bernoulli = BernoulliNB()

# Создаём Multinomial Classifier
model_multinomial = MultinomialNB()

print("For digits dataset")
print("In Gaussian model:", cross_val_score(model_gaussian, X_d, y_d).mean())
print("In Bernoulli model:", cross_val_score(model_bernoulli, X_d, y_d).mean())
print("In Multinomial model:", cross_val_score(model_multinomial, X_d, y_d).mean())

print("")
print("For breast_cancer dataset")
print("In Gaussian model:", cross_val_score(model_gaussian, X_bc, y_bc).mean())
print("In Bernoulli model:", cross_val_score(model_bernoulli, X_bc, y_bc).mean())
print("In Multinomial model:", cross_val_score(model_multinomial, X_bc, y_bc).mean())


For digits dataset
In Gaussian model: 0.818600380355
In Bernoulli model: 0.825823650778
In Multinomial model: 0.870877148974

For breast_cancer dataset
In Gaussian model: 0.936749280609
In Bernoulli model: 0.627420402859
In Multinomial model: 0.894579040193


#### Ответы на вопросы.

1) Каким получилось максимальное качество классификации на датасете breast_cancer?

    На датасете breast_cancer максимальным оказалось качество в модели Gaussian: 0.936749280609.

2) Каким получилось максимальное качество классификации на датасете digits?

    На датасете digits максимальным оказалось качество в модели Multinomial: 0.870877148974.

3) Какие утверждения из приведенных ниже верны?

    (a) На вещественных признаках лучше всего сработал наивный байесовский классификатор с распределением Бернулли.
    Ответ: это утверждение неверно.

    (b) На вещественных признаках лучше всего сработал наивный байесовский классификатор с мультиномиальным распределением.
    Ответ: это утверждение неверно.

    (c) Мультиномиальное распределение лучше показало себя на выборке с целыми неотрицательными значениями признаков.
    Ответ: это утверждение неверно. Мультиномиальное распределение лучше всего показало себя на выборке с вещественными значениями признаков, но, если рассматривать целые неотрицательные значения признаков, то в этом случае мультиномиальное распределение оказалось лучше остальных.

    (d) На вещественных признаках лучше всего сработало нормальное распределение.
    Ответ: это утверждение верно.


##### Вывод:
Считается, что модель Bernoulli является наиболее эффективной для бинарных свойств. Ни в одном из рассмотренных случаев бинарных свойств не оказалось, поэтому, как и ожидалось, эта модель не является самой эффективной в данных примерах.

Модель Multinomial считается наиболее эффективной для выборки с целочисленными значениями признаков. То есть ожидания оправдались.

Модель Gaussian считается наиболее эффективной для выборки с нормально распределёнными вещественными значениями свойств. В рассмотренном примере с вещественными свойствами эта модель оказалась наиболее эффективной.